src/hotspot/os/linux/cgroupUtil_linux.cpp

```diff
@@ -46,3 +47,99 @@ int CgroupUtil::processor_count(CgroupCpuController* cpu_ctrl, int host_cpus) {
   log_trace(os, container)("OSContainer::active_processor_count: %d", result);
   return result;
 }
+
+CgroupMemoryController* CgroupUtil::adjust_controller(CgroupMemoryController* mem) {
+  if (!mem->needs_hierarchy_adjustment()) {
+    // nothing to do
+    return mem;
+  }
+  log_trace(os, container)("Adjusting controller path for memory: %s", mem->subsystem_path());
+  assert(mem->cgroup_path() != nullptr, "invariant");
+  char* orig = os::strdup(mem->cgroup_path());
+  char* cg_path = os::strdup(orig);
+  char* last_slash;
+  julong phys_mem = os::Linux::physical_memory();
+  jlong limit = mem->read_memory_limit_in_bytes(phys_mem);
+  bool path_iterated = false;
+  while (limit < 0 && (last_slash = strrchr(cg_path, '/')) != cg_path) {
+    *last_slash = '\0'; // strip path
+    // update to shortened path and try again
+    mem->set_subsystem_path(cg_path);
+    limit = mem->read_memory_limit_in_bytes(phys_mem);
+    path_iterated = true;
+    if (limit > 0) {
+      log_trace(os, container)("Adjusted controller path for memory to: %s", mem->subsystem_path());
+      os::free(cg_path);
+      os::free(orig);
+      return mem;
+    }
+  }
+  // no lower limit found or limit at leaf
+  os::free(cg_path);
+  if (path_iterated) {
+    mem->set_subsystem_path((char*)"/");
```

Comment:
set_subsystem_path should really take a const char*, and this cast should not be necessary.

Reply:
Done.